flaky test `test_basic_directory_configs_end_to_end`

**log**
```
test_basic_directory_configs_end_to_end(basic_failing_job) (test.functional.test_cluster_basic.TestClusterBasic) ... [32m[2015-06-17 00:19:29.131] 2538 DEBUG   MasterTornadoTh log             
**************************************************
 ClusterRunner
  * Version: 0.5.174-3fd28d9
  * PID:     2538
**************************************************

[0m[36m[2015-06-17 00:19:29.156] 2538 INFO    MasterTornadoTh subcommand      Master service is running on testing-worker-linux-bb3654bb-2-24215-linux-14-67120995:43000.
[0m[32m[2015-06-17 00:19:29.947] 2550 DEBUG   SlaveTornadoThr log             
**************************************************
 ClusterRunner
  * Version: 0.5.174-3fd28d9
  * PID:     2550
**************************************************

[0m[36m[2015-06-17 00:19:29.983] 2538 INFO    MasterTornadoTh cluster_master  Slave on testing-worker-linux-bb3654bb-2-24215-linux-14-67120995:43001 connected to master with 1 executors. (id: 1)
[0m[36m[2015-06-17 00:19:29.985] 2550 INFO    SlaveTornadoThr cluster_slave   Slave testing-worker-linux-bb3654bb-2-24215-linux-14-67120995:43001 connected to master on localhost:43000.
[0m[32m[2015-06-17 00:19:30.101] 2538 DEBUG   MasterTornadoTh directory       Project directory is /tmp/tmpn6x2w0u3
[0m[36m[2015-06-17 00:19:30.102] 2538 INFO    MasterTornadoTh event_log       Queued request for build 1.
[0m[36m[2015-06-17 00:19:30.106] 2538 INFO    RequestHandlerL build_request_h Creating project lock [/tmp/tmpn6x2w0u3] for build 1
[0m[36m[2015-06-17 00:19:30.107] 2538 INFO    Bld1-Preparatio build_request_h Build 1 is waiting for the project lock
[0m[36m[2015-06-17 00:19:30.107] 2538 INFO    Bld1-Preparatio build_request_h Build 1 has acquired project lock
[0m[36m[2015-06-17 00:19:30.108] 2538 INFO    Bld1-Preparatio event_log       Build preparation loop is handling request for build 1.
[0m[36m[2015-06-17 00:19:30.108] 2538 INFO    Bld1-Preparatio build_request_h Fetching project for build 1.
[0m[36m[2015-06-17 00:19:30.108] 2538 INFO    Bld1-Preparatio directory       Build setup complete.
[0m[36m[2015-06-17 00:19:30.109] 2538 INFO    Bld1-Preparatio build_request_h Successfully fetched project for build 1.
[0m[32m[2015-06-17 00:19:30.114] 2538 DEBUG   Bld1-Preparatio directory       Executing command in project: export PROJECT_DIR="/tmp/tmpn6x2w0u3"; seq 0 4 | xargs -I {} echo "This is atom {}"
[0m[32m[2015-06-17 00:19:30.196] 2538 DEBUG   Bld1-Preparatio directory       Command completed with exit code 0.
[0m[36m[2015-06-17 00:19:30.197] 2538 INFO    Bld1-Preparatio event_log       Build 1 successfully prepared and waiting for slaves.
[0m[36m[2015-06-17 00:19:30.229] 2538 INFO    SlaveAllocation slave_allocator Allocating slave testing-worker-linux-bb3654bb-2-24215-linux-14-67120995:43001 to build 1.
[0m[36m[2015-06-17 00:19:30.234] 2550 INFO    SlaveTornadoThr cluster_slave   Executing setup for build 1 (type: directory).
[0m[32m[2015-06-17 00:19:30.234] 2550 DEBUG   SlaveTornadoThr directory       Project directory is /tmp/tmpn6x2w0u3
[0m[36m[2015-06-17 00:19:30.235] 2550 INFO    Bld1-Setup      directory       Build setup complete.
[0m[32m[2015-06-17 00:19:30.235] 2550 DEBUG   Bld1-Setup      directory       Project directory is /tmp/tmpn6x2w0u3
[0m[36m[2015-06-17 00:19:30.242] 2550 INFO    Bld1-Setup      cluster_slave   Notifying master that build setup is complete for build 1.
[0m[1m[31m[2015-06-17 00:19:30.247] 2538 ERROR   MasterTornadoTh cluster_base_ha Exception occurred during request to /v1/slave/1.
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/tornado/web.py", line 1346, in _when_complete
    callback()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/tornado/web.py", line 1367, in _execute_method
    self._when_complete(method(*self.path_args, **self.path_kwargs),
  File "/home/travis/build/box/ClusterRunner/app/util/decorators.py", line 63, in function_with_auth
    return function(self, *args, **kwargs)
  File "/home/travis/build/box/ClusterRunner/app/web_framework/cluster_master_application.py", line 244, in put
    self._cluster_master.handle_slave_state_update(slave, new_slave_state)
  File "/home/travis/build/box/ClusterRunner/app/master/cluster_master.py", line 173, in handle_slave_state_update
    do_transition(slave)
  File "/home/travis/build/box/ClusterRunner/app/master/cluster_master.py", line 212, in _handle_setup_success_on_slave
    build = self.get_build(slave.current_build_id)
  File "/home/travis/build/box/ClusterRunner/app/master/cluster_master.py", line 296, in get_build
    raise ItemNotFoundError('Invalid build id.')
app.util.exceptions.ItemNotFoundError: Invalid build id.
[0m[1m[33m[2015-06-17 00:19:30.308] 2538 WARNING MasterTornadoTh tornado.access  404 PUT /v1/slave/1 (127.0.0.1) 61.98ms
[0m[1m[31m[2015-06-17 00:19:30.309] 2550 ERROR   Bld1-Setup      unhandled_excep Unhandled exception handler caught exception.
Traceback (most recent call last):
  File "/home/travis/build/box/ClusterRunner/app/util/safe_thread.py", line 18, in run
    super().run()
  File "/opt/python/3.4.2/lib/python3.4/threading.py", line 869, in run
    self._target(*self._args, **self._kwargs)
  File "/home/travis/build/box/ClusterRunner/app/slave/cluster_slave.py", line 133, in _async_setup_build
    self._notify_master_of_state_change(SlaveState.SETUP_COMPLETED)
  File "/home/travis/build/box/ClusterRunner/app/slave/cluster_slave.py", line 322, in _notify_master_of_state_change
    secret=Secret.get(), error_on_failure=True)
  File "/home/travis/build/box/ClusterRunner/app/util/network.py", line 83, in put_with_digest
    error_on_failure=error_on_failure)
  File "/home/travis/build/box/ClusterRunner/app/util/decorators.py", line 38, in function_with_retries
    return function(*args, **kwargs)
  File "/home/travis/build/box/ClusterRunner/app/util/network.py", line 71, in put
    return self._request('PUT', *args, **kwargs)
  File "/home/travis/build/box/ClusterRunner/app/util/network.py", line 125, in _request
    format(url, str(resp.status_code), resp.text))
app.util.network._RequestFailedError: Request to http://localhost:43000/v1/slave/1 failed with status_code 404 and response "{"error": "Invalid build id."}"
[0m[32m[2015-06-17 00:19:30.312] 2550 DEBUG   Bld1-Setup      unhandled_excep Executing teardown callback: <bound method ClusterSlave._disconnect_from_master of <app.slave.cluster_slave.ClusterSlave object at 0x2b4c875d4208>>
[0m[36m[2015-06-17 00:19:30.317] 2550 INFO    Bld1-Setup      cluster_slave   Notifying master that this slave is disconnecting.
[0m[36m[2015-06-17 00:19:30.322] 2538 INFO    MasterTornadoTh cluster_master  Slave on testing-worker-linux-bb3654bb-2-24215-linux-14-67120995:43001 was disconnected. (id: 1)
[0m[32m[2015-06-17 00:19:30.332] 2550 DEBUG   Bld1-Setup      unhandled_excep Executing teardown callback: <bound method ClusterSlave._do_build_teardown_and_reset of <app.slave.cluster_slave.ClusterSlave object at 0x2b4c875d4208>>
[0m[36m[2015-06-17 00:19:30.346] 2550 INFO    Bld1-Setup      cluster_slave   Executing teardown for build 1.
[0m[36m[2015-06-17 00:19:30.354] 2550 INFO    Bld1-Setup      directory       ProjectType teardown complete.
[0m[36m[2015-06-17 00:19:30.354] 2550 INFO    Bld1-Setup      cluster_slave   Build teardown complete for build 1.
[0m[32m[2015-06-17 00:19:30.354] 2550 DEBUG   Bld1-Setup      unhandled_excep Executing teardown callback: <function ServiceSubcommand._write_pid_file.<locals>.remove_pid_file at 0x2b4c87813e18>
[0m[32m[2015-06-17 00:19:30.355] 2550 DEBUG   Bld1-Setup      unhandled_excep Executing teardown callback: functools.partial(<bound method EPollIOLoop.add_callback of <tornado.platform.epoll.EPollIOLoop object at 0x2b4c875d4c50>>, callback=<bound method EPollIOLoop.stop of <tornado.platform.epoll.EPollIOLoop object at 0x2b4c875d4c50>>)
[0m[33m[2015-06-17 00:19:30.356] 2550 NOTICE  SlaveTornadoThr subcommand      Slave server was stopped.
[0m[36m[2015-06-17 00:19:40.166] 2334 INFO    MainThread      functional_test Waiting on build queue to become empty.
[0m[36m[2015-06-17 00:19:41.174] 2334 INFO    MainThread      functional_test Waiting on build queue to become empty.
[0m[36m[2015-06-17 00:19:42.183] 2334 INFO    MainThread      functional_test Waiting on build queue to become empty.
[0m[36m[2015-06-17 00:19:43.196] 2334 INFO    MainThread      functional_test Waiting on build queue to become empty.
[0m[36m[2015-06-17 00:19:44.203] 2334 INFO    MainThread      functional_test Waiting on build queue to become empty.
[0m[1m[31m[2015-06-17 00:19:45.205] 2334 ERROR   MainThread      functional_test Master queue did not become empty before timeout.
[0m[36m[2015-06-17 00:19:45.205] 2538 INFO    MainThread      unhandled_excep SIGTERM signal received. Triggering teardown.
[0m[32m[2015-06-17 00:19:45.206] 2538 DEBUG   MainThread      unhandled_excep Executing teardown callback: <function ServiceSubcommand._write_pid_file.<locals>.remove_pid_file at 0x2ae2e4213d08>
[0m[32m[2015-06-17 00:19:45.207] 2538 DEBUG   MainThread      unhandled_excep Executing teardown callback: functools.partial(<bound method EPollIOLoop.add_callback of <tornado.platform.epoll.EPollIOLoop object at 0x2ae2db7f4d68>>, callback=<bound method EPollIOLoop.stop of <tornado.platform.epoll.EPollIOLoop object at 0x2ae2db7f4d68>>)
[0m[33m[2015-06-17 00:19:45.208] 2538 NOTICE  MasterTornadoTh subcommand      Master server was stopped.
[0mERROR
FAIL
```

**test report**
```
======================================================================
ERROR: test_basic_directory_configs_end_to_end(basic_failing_job) (test.functional.test_cluster_basic.TestClusterBasic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 493, in urlopen
    body=body, headers=headers)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 291, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/opt/python/3.4.2/lib/python3.4/http/client.py", line 1090, in request
    self._send_request(method, url, body, headers)
  File "/opt/python/3.4.2/lib/python3.4/http/client.py", line 1128, in _send_request
    self.endheaders(body)
  File "/opt/python/3.4.2/lib/python3.4/http/client.py", line 1086, in endheaders
    self._send_output(message_body)
  File "/opt/python/3.4.2/lib/python3.4/http/client.py", line 924, in _send_output
    self.send(msg)
  File "/opt/python/3.4.2/lib/python3.4/http/client.py", line 859, in send
    self.connect()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/packages/urllib3/connection.py", line 106, in connect
    conn = self._new_conn()
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/packages/urllib3/connection.py", line 90, in _new_conn
    (self.host, self.port), self.timeout, *extra_args)
  File "/opt/python/3.4.2/lib/python3.4/socket.py", line 509, in create_connection
    raise err
  File "/opt/python/3.4.2/lib/python3.4/socket.py", line 500, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/adapters.py", line 327, in send
    timeout=timeout
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 543, in urlopen
    raise MaxRetryError(self, url, e)
requests.packages.urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=43001): Max retries exceeded with url: /v1 (Caused by <class 'ConnectionRefusedError'>: [Errno 111] Connection refused)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/genty/genty.py", line 279, in <lambda>
    *(my_self + dataset)
  File "/home/travis/build/box/ClusterRunner/test/functional/test_cluster_basic.py", line 30, in test_basic_directory_configs_end_to_end
    slave.block_until_idle(timeout=5)  # ensure slave teardown has finished before making assertions
  File "/home/travis/build/box/ClusterRunner/app/client/cluster_api_client.py", line 218, in block_until_idle
    poll.wait_for(is_slave_idle, timeout_seconds=timeout)
  File "/home/travis/build/box/ClusterRunner/app/util/poll.py", line 25, in wait_for
    if boolean_predicate():
  File "/home/travis/build/box/ClusterRunner/app/client/cluster_api_client.py", line 215, in is_slave_idle
    response_data = self.get_slave_status()
  File "/home/travis/build/box/ClusterRunner/app/client/cluster_api_client.py", line 225, in get_slave_status
    response_data = self._network.get(slave_status_url).json()
  File "/home/travis/build/box/ClusterRunner/app/util/network.py", line 37, in get
    return self._request('GET', *args, **kwargs)
  File "/home/travis/build/box/ClusterRunner/app/util/network.py", line 122, in _request
    resp = self._session.request(method, url, data=data_to_send, *args, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/sessions.py", line 456, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/sessions.py", line 559, in send
    r = adapter.send(request, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/requests/adapters.py", line 375, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=43001): Max retries exceeded with url: /v1 (Caused by <class 'ConnectionRefusedError'>: [Errno 111] Connection refused)

======================================================================
FAIL: test_basic_directory_configs_end_to_end(basic_failing_job) (test.functional.test_cluster_basic.TestClusterBasic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/box/ClusterRunner/test/framework/functional/base_functional_test_case.py", line 76, in tearDown
    'with code {}.'.format(service.url, service.return_code))
AssertionError: 1 != 0 : Service running on url: http://localhost:43001 should exit with code 0, but exited with code 1.

----------------------------------------------------------------------
Ran 8 tests in 65.275s
```

--- a/app/master/slave.py
+++ b/app/master/slave.py
@@ -70,8 +70,9 @@ class Slave(object):
             'project_type_params': slave_project_type_params,
             'build_executor_start_index': build.num_executors_allocated,
         }
-        self._network.post_with_digest(setup_url, post_data, Secret.get())
+
         self.current_build_id = build.build_id()
+        self._network.post_with_digest(setup_url, post_data, Secret.get())
 
     def teardown(self):
         """

--- a/test/unit/master/test_slave.py
+++ b/test/unit/master/test_slave.py
@@ -65,6 +65,23 @@ class TestSlave(BaseUnitTestCase):
             Secret.get()
         )
 
+    def test_build_id_is_set_on_master_before_telling_slave_to_setup(self):
+        # This test enforces an ordering that avoids a race where the slave finishes setup and posts back before the
+        # master has actually set the slave's current_build_id.
+        slave = self._create_slave()
+        mock_build = Mock()
+
+        def assert_slave_build_id_is_already_set(*args, **kwargs):
+            self.assertEqual(slave.current_build_id, mock_build.build_id(),
+                             'slave.current_build_id should be set before the master tells the slave to do setup.')
+
+        slave._network.post_with_digest = Mock(side_effect=assert_slave_build_id_is_already_set)
+        slave.setup(mock_build)
+
+        self.assertEqual(slave._network.post_with_digest.call_count, 1,
+                         'The behavior that this test is checking depends on slave setup being triggered via '
+                         'slave._network.post_with_digest().')
+
     def test_is_alive_returns_cached_value_if_use_cache_is_true(self):
         slave = self._create_slave()
         slave._is_alive = False

